src/include/kompute/Core.hpp

```diff
@@ -32,53 +32,76 @@ static const char* KOMPUTE_LOG_TAG = "KomputeLog";
 #endif
 #endif
 
+#if defined(KOMPUTE_BUILD_PYTHON)
+    #include <pybind11/pybind11.h>
+    namespace py = pybind11;
+    //from python/src/main.cpp
+    extern py::object kp_logger;
+#endif
+
+
 #ifndef KOMPUTE_LOG_OVERRIDE
-#if KOMPUTE_ENABLE_SPDLOG
-#include <spdlog/spdlog.h>
-#else
-#include <iostream>
-#if SPDLOG_ACTIVE_LEVEL > 1
-#define SPDLOG_DEBUG(message, ...)
-#else
-#if defined(VK_USE_PLATFORM_ANDROID_KHR)
-#define SPDLOG_DEBUG(message, ...)                                             \
-    ((void)__android_log_print(ANDROID_LOG_DEBUG, KOMPUTE_LOG_TAG, message))
-#else
-#define SPDLOG_DEBUG(message, ...)                                             \
-    std::cout << "DEBUG: " << message << std::endl
-#endif // VK_USE_PLATFORM_ANDROID_KHR
-#endif // SPDLOG_ACTIVE_LEVEL > 1
-#if SPDLOG_ACTIVE_LEVEL > 2
-#define SPDLOG_INFO(message, ...)
-#else
-#if defined(VK_USE_PLATFORM_ANDROID_KHR)
-#define SPDLOG_INFO(message, ...)                                              \
-    ((void)__android_log_print(ANDROID_LOG_INFO, KOMPUTE_LOG_TAG, message))
-#else
-#define SPDLOG_INFO(message, ...) std::cout << "INFO: " << message << std::endl
-#endif // VK_USE_PLATFORM_ANDROID_KHR
-#endif // SPDLOG_ACTIVE_LEVEL > 2
-#if SPDLOG_ACTIVE_LEVEL > 3
-#define SPDLOG_WARN(message, ...)
-#else
-#if defined(VK_USE_PLATFORM_ANDROID_KHR)
-#define SPDLOG_WARN(message, ...)                                              \
-    ((void)__android_log_print(ANDROID_LOG_INFO, KOMPUTE_LOG_TAG, message))
-#else
-#define SPDLOG_WARN(message, ...)                                              \
-    std::cout << "WARNING: " << message << std::endl
-#endif // VK_USE_PLATFORM_ANDROID_KHR
-#endif // SPDLOG_ACTIVE_LEVEL > 3
-#if SPDLOG_ACTIVE_LEVEL > 4
-#define SPDLOG_ERROR(message, ...)
-#else
-#if defined(VK_USE_PLATFORM_ANDROID_KHR)
-#define SPDLOG_ERROR(message, ...)                                             \
-    ((void)__android_log_print(ANDROID_LOG_INFO, KOMPUTE_LOG_TAG, message))
-#else
-#define SPDLOG_ERROR(message, ...)                                             \
-    std::cout << "ERROR: " << message << std::endl
-#endif // VK_USE_PLATFORM_ANDROID_KHR
-#endif // SPDLOG_ACTIVE_LEVEL > 4
-#endif // KOMPUTE_SPDLOG_ENABLED
+    #if KOMPUTE_ENABLE_SPDLOG
+        #include <spdlog/spdlog.h>
+    #else
+        #include <iostream>
+        #if SPDLOG_ACTIVE_LEVEL > 1
+            #define SPDLOG_DEBUG(message, ...)
+        #else
+            #if defined(VK_USE_PLATFORM_ANDROID_KHR)
+                #define SPDLOG_DEBUG(message, ...)                                             \
+                ((void)__android_log_print(ANDROID_LOG_DEBUG, KOMPUTE_LOG_TAG, message))
+            #elif defined(KOMPUTE_BUILD_PYTHON)
+                #define SPDLOG_DEBUG(message, ...)                                             \
+                kp_logger.attr("debug")(message);
```

Comment:
Just to confirm, to the current knowledge (there may be some nuances) this function isn't expected to throw exceptions that should be caught right?

Reply:
To be honest, I don't know. Will investigate

Reply:
The `.attr()` function can fail if the user does something stupid like `del logging.getLogger('kp').__class__.warning`. So I'm caching the individual logging functions now. Should also improve the performance a little bit.

Reply:
Interesting, yeah I doubt people would do that, but I agree with Python you also never know - looks good